### PR TITLE
BaSP-TG-10: ADD Get, GetById, Create for Task

### PR DESCRIPTION
**Pull request by fede230298**
- New model for Tasks
- Router implementation for Tasks
- Add GET to route '/tasks' to get a complete list of Tasks
- Add GET to route '/tasks/:id' to get a Task filtered by its ID
- Add POST to route '/tasks' to create a new Task
- Added validations with Joi as a middleware
- Updated Postman collection

### CODE REVIEW

File: src/controllers/tasks.js [hunk 1]
```diff
@@ -1,24 +1,67 @@
...
+    return res.status(200).json({
+      message: 'Tasks found',
+      data: tasks,
+      error: false,
     });
-  } else {
-    res.status(404).json({
-      message: 'Task not found',
+  } catch (error) {
+    return res.json({
```
Comment (GinaSchP) [inline src/controllers/tasks.js:16]: The status of the error is missing, must be:
 return res.status(400).json({
               message: error, // o error.toString()
               data: undefined,
               error: true
           });

File: src/controllers/tasks.js [hunk 2]
```diff
@@ -1,24 +1,67 @@
...
+    const { id } = req.params;
+    const tasks = await Tasks.findById(id);
+
+    return res.status(200).json({
+      message: 'Tasks found',
+      data: tasks,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
```
Comment (GinaSchP) [inline src/controllers/tasks.js:34]: same as above

File: src/controllers/tasks.js [hunk 3]
```diff
@@ -1,24 +1,67 @@
...
 
-export const getTaskById = (req, res) => {
-  const taskId = parseInt(req.params.id, 10);
-  const foundTask = tasks.find((task) => task.id === taskId);
-  if (foundTask) {
-    res.status(200).json({
-      data: foundTask,
+export const getTaskList = async (req, res) => {
+  try {
+    const tasks = await Tasks.find();
```
Comment (GinaSchP) [inline src/controllers/tasks.js:8]: Here you can implement Tasks.find(req.query), and if there is a filter, it is applied directly.

**Review by GinaSchP — CHANGES_REQUESTED** [review]

File: src/controllers/tasks.js [hunk 4]
```diff
@@ -1,24 +1,67 @@
...
+      message: 'An error occurred',
+      error,
+    });
+  }
+};
+
+export const getTaskById = async (req, res) => {
+  try {
+    const { id } = req.params;
+    const tasks = await Tasks.findById(id);
```
Comment (GinaSchP) [inline src/controllers/tasks.js:26]: "tasks" should be "task" Also, you must validate if the method returns something (with an if statement) and if not return a 404 error with a not found message.

**Review by GinaSchP — CHANGES_REQUESTED** [review]

File: src/routes/index.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const express = require('express');
+
+const router = express.Router();
+const tasksRouter = require('./tasks');
```
Comment (lmalvasia) [inline src/routes/index.js:4]: All the imports should be at the beginning of the project. We also have to use the "import" (ES6).

File: src/routes/tasks.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+const express = require('express');
```
Comment (lmalvasia) [inline src/routes/tasks.js:1]: use "import" to import the files and libraries

File: src/routes/tasks.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+const router = express.Router();
+
+router
+  .get('/', getTaskList)
+  .get('/:id', getTaskById)
+  .delete('/:id', deleteTaskById)
+  .post('/', taskValidation.validateCreation, createNewTask)
+  .put('/:id', editTask);
+
+module.exports = router;
```
Comment (lmalvasia) [inline src/routes/tasks.js:21]: export router

File: src/validations/tasks.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import Joi from 'joi';
+
+export const validateCreation = (req, res, next) => {
```
Comment (lmalvasia) [inline src/validations/tasks.js:3]:
```suggestion
const validateTaskBody = (req, res, next) => {
```

File: src/validations/tasks.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      message: `There was an error: ${validation.error.details[0].message}`,
+      data: undefined,
+      error: true,
+    });
+  }
+
+  return next();
+};
+
+export default {
```
Comment (lmalvasia) [inline src/validations/tasks.js:21]: export default validateTaskBody;

File: src/controllers/tasks.js [hunk 10]
```diff
@@ -1,24 +1,77 @@
...
+    });
+  }
+};
+
+export const getTaskById = async (req, res) => {
+  try {
+    const { id } = req.params;
+    const task = await Tasks.findById(id);
+
+    return res.status(200).json({
```
Comment (lmalvasia) [inline src/controllers/tasks.js:36]: Add a validation to return a different message when "task" is undefined (not exist).

File: src/controllers/tasks.js [hunk 11]
```diff
@@ -1,24 +1,83 @@
...
-  const taskId = parseInt(req.params.id, 10);
-  const foundTask = tasks.find((task) => task.id === taskId);
-  if (foundTask) {
-    res.status(200).json({
-      data: foundTask,
+export const getTaskList = async (req, res) => {
+  try {
+    const tasks = await Tasks.find(req.query);
+
+    if (tasks.length === 0) {
```
Comment (lmalvasia) [inline src/controllers/tasks.js:9]:
```suggestion
    if (!tasks.length) {
```

File: src/controllers/tasks.js [hunk 12]
```diff
@@ -1,24 +1,83 @@
...
+    });
+  }
+};
+
+export const getTaskById = async (req, res) => {
+  try {
+    const { id } = req.params;
+    const task = await Tasks.findById(id);
+
+    if (task === null) {
```
Comment (lmalvasia) [inline src/controllers/tasks.js:35]:
```suggestion
    if (!task) {
```

File: src/validations/tasks.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      message: `There was an error: ${validation.error.details[0].message}`,
+      data: undefined,
+      error: true,
+    });
+  }
+
+  return next();
+};
+
+export const validateQuery = (req, res, next) => {
```
Comment (lmalvasia) [inline src/validations/tasks.js:21]:
```suggestion
export const validateTaskQueryParams = (req, res, next) => {
```

**Review by lmalvasia — APPROVED** [review]

**Review by GinaSchP — DISMISSED** [review]

**Review by GinaSchP — APPROVED** [review]

**Pull request merged**